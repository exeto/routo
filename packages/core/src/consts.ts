export const NOT_FOUND = '@@routo/NOT_FOUND';
