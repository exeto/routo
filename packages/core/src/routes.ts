import { pathToRegexp, compile, Key } from 'path-to-regexp';

import { Route, ExtendedRoute, RouteStorage } from './types';

export const createRoute = (route: Route): ExtendedRoute => {
  const keys: Key[] = [];
  const regexp = pathToRegexp(route.path, keys, { strict: true });

  return {
    ...route,
    regexp,
    keys,
    createPathname: compile(route.path),
  };
};

export const createRouteStorage = (
  notFoundRoute: Route,
  routes: Route[],
): RouteStorage => {
  const extendedRoutes = [...routes, notFoundRoute].map(createRoute);
  const extendedNotFoundRoute = extendedRoutes[extendedRoutes.length - 1];

  return {
    getById(id) {
      return (
        extendedRoutes.find((route) => id === route.id) || extendedNotFoundRoute
      );
    },

    getByPathname(pathname) {
      return (
        extendedRoutes.find(({ regexp }) => regexp.test(pathname)) ||
        extendedNotFoundRoute
      );
    },
  };
};
