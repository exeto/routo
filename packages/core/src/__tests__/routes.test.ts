import { NOT_FOUND } from '../consts';
import { createRoute, createRouteStorage } from '../routes';

const notFoundRoute = {
  id: NOT_FOUND,
  path: '/404',
};

const HOME = 'router/HOME';

const POST = 'router/POST';

const routes = [
  {
    id: HOME,
    path: '/',
  },
  {
    id: POST,
    path: '/posts/:id',
  },
];

describe('createRoute', () => {
  it('should extend simple route', () => {
    const extendedRoute = createRoute(routes[0]);

    expect(extendedRoute).toMatchSnapshot();
  });

  it('should extend parameterized route', () => {
    const extendedRoute = createRoute(routes[1]);

    expect(extendedRoute).toMatchSnapshot();
  });

  describe('createPathname', () => {
    it('should return pathname for simple route', () => {
      const extendedRoute = createRoute(routes[0]);

      expect(extendedRoute.createPathname({})).toBe('/');
    });

    it('should return pathname for parameterized route', () => {
      const extendedRoute = createRoute(routes[1]);

      expect(extendedRoute.createPathname({ id: 42 })).toBe('/posts/42');
    });
  });
});

describe('RouteStorage instance', () => {
  const routeStorage = createRouteStorage(notFoundRoute, routes);

  describe('getById', () => {
    it('should return not found route', () => {
      expect(routeStorage.getById('router/SOMETHING')).toMatchSnapshot();
      expect(routeStorage.getById(NOT_FOUND)).toMatchSnapshot();
    });

    it('should return simple route', () => {
      expect(routeStorage.getById(HOME)).toMatchSnapshot();
    });

    it('should return parameterized route', () => {
      expect(routeStorage.getById(POST)).toMatchSnapshot();
    });
  });

  describe('getByPathname', () => {
    it('should return not found route', () => {
      expect(routeStorage.getByPathname('/something')).toMatchSnapshot();
      expect(routeStorage.getByPathname('/something/42')).toMatchSnapshot();
      expect(routeStorage.getByPathname('/posts')).toMatchSnapshot();
    });

    it('should return simple route', () => {
      expect(routeStorage.getByPathname('/')).toMatchSnapshot();
    });

    it('should return parameterized route', () => {
      expect(routeStorage.getByPathname('/posts/42')).toMatchSnapshot();
    });
  });
});
