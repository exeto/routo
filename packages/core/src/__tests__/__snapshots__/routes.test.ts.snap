// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouteStorage instance getById should return not found route 1`] = `
Object {
  "createPathname": [Function],
  "id": "@@routo/NOT_FOUND",
  "keys": Array [],
  "path": "/404",
  "regexp": /\\^\\\\/404\\$/i,
}
`;

exports[`RouteStorage instance getById should return not found route 2`] = `
Object {
  "createPathname": [Function],
  "id": "@@routo/NOT_FOUND",
  "keys": Array [],
  "path": "/404",
  "regexp": /\\^\\\\/404\\$/i,
}
`;

exports[`RouteStorage instance getById should return parameterized route 1`] = `
Object {
  "createPathname": [Function],
  "id": "router/POST",
  "keys": Array [
    Object {
      "modifier": "",
      "name": "id",
      "pattern": "[^\\\\/#\\\\?]+?",
      "prefix": "/",
      "suffix": "",
    },
  ],
  "path": "/posts/:id",
  "regexp": /\\^\\\\/posts\\(\\?:\\\\/\\(\\[\\^\\\\/#\\\\\\?\\]\\+\\?\\)\\)\\$/i,
}
`;

exports[`RouteStorage instance getById should return simple route 1`] = `
Object {
  "createPathname": [Function],
  "id": "router/HOME",
  "keys": Array [],
  "path": "/",
  "regexp": /\\^\\\\/\\$/i,
}
`;

exports[`RouteStorage instance getByPathname should return not found route 1`] = `
Object {
  "createPathname": [Function],
  "id": "@@routo/NOT_FOUND",
  "keys": Array [],
  "path": "/404",
  "regexp": /\\^\\\\/404\\$/i,
}
`;

exports[`RouteStorage instance getByPathname should return not found route 2`] = `
Object {
  "createPathname": [Function],
  "id": "@@routo/NOT_FOUND",
  "keys": Array [],
  "path": "/404",
  "regexp": /\\^\\\\/404\\$/i,
}
`;

exports[`RouteStorage instance getByPathname should return not found route 3`] = `
Object {
  "createPathname": [Function],
  "id": "@@routo/NOT_FOUND",
  "keys": Array [],
  "path": "/404",
  "regexp": /\\^\\\\/404\\$/i,
}
`;

exports[`RouteStorage instance getByPathname should return parameterized route 1`] = `
Object {
  "createPathname": [Function],
  "id": "router/POST",
  "keys": Array [
    Object {
      "modifier": "",
      "name": "id",
      "pattern": "[^\\\\/#\\\\?]+?",
      "prefix": "/",
      "suffix": "",
    },
  ],
  "path": "/posts/:id",
  "regexp": /\\^\\\\/posts\\(\\?:\\\\/\\(\\[\\^\\\\/#\\\\\\?\\]\\+\\?\\)\\)\\$/i,
}
`;

exports[`RouteStorage instance getByPathname should return simple route 1`] = `
Object {
  "createPathname": [Function],
  "id": "router/HOME",
  "keys": Array [],
  "path": "/",
  "regexp": /\\^\\\\/\\$/i,
}
`;

exports[`createRoute should extend parameterized route 1`] = `
Object {
  "createPathname": [Function],
  "id": "router/POST",
  "keys": Array [
    Object {
      "modifier": "",
      "name": "id",
      "pattern": "[^\\\\/#\\\\?]+?",
      "prefix": "/",
      "suffix": "",
    },
  ],
  "path": "/posts/:id",
  "regexp": /\\^\\\\/posts\\(\\?:\\\\/\\(\\[\\^\\\\/#\\\\\\?\\]\\+\\?\\)\\)\\$/i,
}
`;

exports[`createRoute should extend simple route 1`] = `
Object {
  "createPathname": [Function],
  "id": "router/HOME",
  "keys": Array [],
  "path": "/",
  "regexp": /\\^\\\\/\\$/i,
}
`;
